public/app/features/dashboard/utils/panelMerge.ts

```diff
@@ -0,0 +1,86 @@
+import { PanelModel as IPanelModel } from '@grafana/data';
+import { isEqual } from 'lodash';
+import { PanelModel } from '../state';
+
+export interface PanelMergeInfo {
+  changed: boolean;
+  panels: PanelModel[];
+  actions: Record<string, number[]>;
+}
+
+export function mergePanels(current: PanelModel[], data: IPanelModel[]): PanelMergeInfo {
+  const panels: PanelModel[] = [];
+  const info = {
+    changed: false,
+    actions: {
+      add: [] as number[],
+      remove: [] as number[],
+      replace: [] as number[],
+      noop: [] as number[],
+    },
+    panels,
+  };
+  const inputPanels = new Map<number, IPanelModel>();
+  for (const p of data) {
+    inputPanels.set(p.id, p);
+  }
+
+  for (const panel of current) {
+    const target = inputPanels.get(panel.id) as PanelModel; // the same model
+    if (!target) {
+      info.changed = true;
+      info.actions.remove.push(panel.id);
+      panel.destroy();
+      continue;
+    }
+    inputPanels.delete(panel.id);
+
+    // Type changed trigger full refresh
+    if (panel.type !== target.type) {
+      panel.destroy();
+
+      const next = new PanelModel(target);
+      next.key = `${next.id}-update-${Date.now()}`; // force react replacement
+      panels.push(next);
+      info.changed = true;
+      info.actions.replace.push(panel.id);
+      continue;
+    }
+
+    // All new properties are the same
+    const save = panel.getSaveModel();
+    let same = true;
+    for (const [key, value] of Object.entries(target)) {
+      if (!isEqual(value, save[key])) {
```

Comment:
`_.isEqual()` is supposed to not care about object key order, but my experiments suggest it does matter --- or something else if happening